File: typescript/src/booking/application/policies/paymentPolicy.ts
// Policy che traduce gli esiti del Payment BC in comandi di conferma/rifiuto del Booking BC.


import { ClientReference } from '@/common/domain/primitive/clientReference.js';
import type { PaymentResultEvent } from '@/payment/domain/events/paymentResultEvents.js';
import type { Policy } from '@/common/application/policy.js';
import { match } from 'ts-pattern';
import { Booking } from '../../domain/booking/booking.js';
import { BookingId } from '../../domain/booking/bookingId.js';
import type { BookingRepository } from '../../domain/ports/bookingRepository.js';
import { requireArgument } from '@/common/utils/requireArgument.js';
import { requireDependency } from '@/common/utils/requireDependency.js';
import {
  type BookingConfirmationCommand,
  confirmBooking,
  rejectBooking,
} from '../../application/commands/bookingConfirmationCommands.js';

export class PaymentPolicy implements Policy<PaymentResultEvent, BookingConfirmationCommand> {
  private readonly _bookingRepository: BookingRepository;

  constructor(bookingRepository: BookingRepository) {
    requireDependency(bookingRepository, "bookingRepository");
    this._bookingRepository = bookingRepository;
  }

  evaluate(event: PaymentResultEvent): BookingConfirmationCommand | null {
    requireArgument(event, 'Payment event');

    return match(event)
      .with({ kind: 'PaymentAccepted' }, (accepted) => {
        const booking = this.findBooking(accepted.clientReference);
        return booking === null
          ? null
          : confirmBooking(
              this.bookingIdFrom(accepted.clientReference),
              booking.giftCardId(),
              accepted.amount
            );
      })
      .with({ kind: 'PaymentRejected' }, (rejected) => {
        const booking = this.findBooking(rejected.clientReference);
        return booking === null
          ? null
          : rejectBooking(
              this.bookingIdFrom(rejected.clientReference),
              booking.giftCardId(),
              rejected.amount
            );
      })
      .with({ kind: 'PaymentExpired' }, () => null)
      .exhaustive();
  }

  private findBooking(clientReference: ClientReference): Booking | null {
    return this._bookingRepository.findById(this.bookingIdFrom(clientReference));
  }

  private bookingIdFrom(clientReference: ClientReference): BookingId {
    return new BookingId(clientReference.value);
  }
}

File: java/src/main/java/it/giannibombelli/wsc2026/booking/application/policies/PaymentPolicy.java
package it.giannibombelli.wsc2026.booking.application.policies;

import it.giannibombelli.wsc2026.common.utils.Require;

import it.giannibombelli.wsc2026.booking.application.commands.BookingConfirmationCommands;
import it.giannibombelli.wsc2026.booking.domain.booking.Booking;
import it.giannibombelli.wsc2026.booking.domain.booking.BookingId;
import it.giannibombelli.wsc2026.booking.domain.ports.BookingRepository;
import it.giannibombelli.wsc2026.common.application.Policy;
import it.giannibombelli.wsc2026.common.domain.primitive.ClientReference;
import it.giannibombelli.wsc2026.payment.domain.events.PaymentResultEvents;

import java.util.UUID;

import static it.giannibombelli.wsc2026.booking.application.commands.BookingConfirmationCommands.ConfirmBooking;
import static it.giannibombelli.wsc2026.booking.application.commands.BookingConfirmationCommands.RejectBooking;
import static it.giannibombelli.wsc2026.payment.domain.events.PaymentResultEvents.*;


public final class PaymentPolicy implements Policy<PaymentResultEvents, BookingConfirmationCommands> {
    private final BookingRepository bookingRepository;

    public PaymentPolicy(BookingRepository bookingRepository) {
        Require.requireDependency(bookingRepository, "bookingRepository");
        this.bookingRepository = bookingRepository;
    }

    @Override
    public BookingConfirmationCommands evaluate(PaymentResultEvents event) {
        Require.requireArgument(event, "Payment event");
        return switch (event) {
            case PaymentAccepted it -> {
                Booking booking = findBooking(it.clientReference());
                yield booking == null
                    ? null
                    : new ConfirmBooking(bookingIdFrom(it.clientReference()), booking.giftCardReference(), it.amount());
            }
            case PaymentRejected it -> {
                Booking booking = findBooking(it.clientReference());
                yield booking == null
                    ? null
                    : new RejectBooking(bookingIdFrom(it.clientReference()), booking.giftCardReference(), it.amount());
            }
            case PaymentExpired it -> null;
        };
    }

    private Booking findBooking(ClientReference clientReference) {
        return bookingRepository.findById(bookingIdFrom(clientReference)).orElse(null);
    }

    private BookingId bookingIdFrom(ClientReference clientReference) {
        return new BookingId(clientReference.value());
    }
}
